### ISSUE TITLE
Update Analytics report deep links to use `getServiceReportURL` selector

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2405]
After #2228, all Analytics frontend report deep links should be updated to rely on the new `getServiceReportURL` selector of the `modules/analytics` store. This will fix various instances in which arguments for those URLs are incorrectly encoded.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* All deep links to Analytics (i.e. links that rely on the `getServiceURL` selector of `modules/analytics` with a URL starting with `/report`) should be updated to rely on the `getServiceReportURL` selector.
    * Parameters that are currently manually "encoded" should be passed to `getServiceReportURL` as `reportArgs`.
    * Those links no longer need to pass the Analytics account, property and view IDs since those are handled internally by that selector, simplifying the surrounding code.

## Implementation Brief
- Using `assets/js/modules/analytics/components/common/AcquisitionPieChart.js`
	- Update the `sourceURI` constant to use the `getServiceReportURL` selector with `url` being the current entity URL.
		```js
		const sourceURI = useSelect(
			( select ) => select( STORE_NAME ).getServiceReportURL( 'trafficsources-overview', { '_r.drilldown': `analytics.pagePath:${ url }` } )
		);

		```
- Using `assets/js/modules/analytics/components/dashboard/AnalyticsDashboardWidget.js`
	- Update the `topContentServiceURL`, `topAcquisitionServiceURL`, `visitorsOverview` variables to use the `getServiceReportURL` selector:
		```js
		const topContentServiceURL = useSelect( ( select ) => select( STORE_NAME ).getServiceReportURL( 'content-pages' ) );
		const topAcquisitionServiceURL = useSelect( ( select ) => select( STORE_NAME ).getServiceReportURL( 'trafficsources-overview' ) );
		const visitorsOverview = useSelect( ( select ) => select( STORE_NAME ).getServiceReportURL( 'visitors-overview' ) );
		```
- Using `assets/js/modules/analytics/components/dashboard/AnalyticsDashboardWidgetTopPagesTable.js`
	- Remove the `baseServiceURL` variable
	- In the `dataMapped` array, set the `link` for the item being looped to `url` only.
		```js
		links[ i ] = url;
		```
	- Add the `PrimaryLink` option to the `options` constant. E.g can be found in `assets/js/modules/analytics/components/dashboard/AnalyticsAdSenseDashboardWidgetTopPagesTable.js`
		```js
		PrimaryLink: withSelect( ( select, { href = '/' } ) => {
			const serviceURL = select( STORE_NAME ).getServiceReportURL( 'content-drilldown', {
				'explorer-table.plotKeys': '[]',
				'_r.drilldown': `analytics.pagePath:${ href }`,
			} );

			return {
				href: serviceURL,
				external: true,
			};
		} )( Link ),
		```
- Using `assets/js/modules/analytics/components/dashboard/DashboardBounceRateWidget.js`
	- Update `serviceURL` to use `getServiceReportURL` with `url` being the current entity URL.
		```js
		serviceURL: store.getServiceReportURL(
			'visitors-overview',
			{
				'_r.drilldown': `analytics.pagePath:${ url }`,
			},
		),
		```
- Using `assets/js/modules/analytics/components/dashboard/DashboardGoalsWidget.js`
	- Update `serviceURL`
		```js
		serviceURL: store.getServiceReportURL( 'conversions-goals-overview' ),
		```
- Using `assets/js/modules/analytics/components/dashboard/DashboardPopularPagesWidget.js`
	- Update `analyticsMainURL`
		```js
		analyticsMainURL: store.getServiceReportURL( 'content-pages' ),
		```
- Using `assets/js/modules/analytics/components/dashboard/DashboardUniqueVisitorsWidget.js`
	- Update `serviceURL` with `url` being the current entity URL.
		```js
			serviceURL: store.getServiceReportURL(
				'visitors-overview',
				{
					'_r.drilldown': `analytics.pagePath:${ url }`,
				},
			),
		```
- Using `assets/js/modules/analytics/components/dashboard/LegacyAnalyticsDashboardWidgetPopularPagesTable.js`
	- Update `serviceURL` to use `getServiceReportURL`
		```js
			select( STORE_NAME ).getServiceReportURL( 'content-pages' );
		```
- Using `assets/js/modules/analytics/components/dashboard/LegacyAnalyticsDashboardWidgetTopLevel.js`
	- Update `uniqueVisitorsServiceURL` and `goalsServiceURL` with `url` being the current entity URL. 
	```js
		const uniqueVisitorsServiceURL = useSelect( ( select ) => select( STORE_NAME ).getServiceReportURL(
			`visitors-overview`,
			{
				'_r.drilldown': `analytics.pagePath:${ url }`,
			},
		) );

		const goalsServiceURL = useSelect( ( select ) => select( STORE_NAME ).getServiceReportURL(
			`conversions-goals-overview`,
			{
				'_r.drilldown': `analytics.pagePath:${ url }`,
			}
		) );
	```
- Using `assets/js/modules/analytics/components/dashboard/LegacyDashboardAcquisitionPieChart.js`
	- Update `sourceURI` with `url` being the current entity URL. 
		```js
		const sourceURI = useSelect( ( select ) => select( STORE_NAME ).getServiceReportURL(
			'trafficsources-overview',
			{
				'_r.drilldown': `analytics.pagePath:${ url }`,
			},
		) );
		```
- Using `assets/js/modules/analytics/components/dashboard-details/AnalyticsDashboardDetailsWidgetTopAcquisitionSources.js`
	- Update `serviceURL` to use `getServiceReportURL` with `url` being the current entity URL. 
	```js
	select( STORE_NAME ).getServiceReportURL( 'trafficsources-overview', { '_r.drilldown': `analytics.pagePath:${ url }` } );
	```
- Remove unused `accountID`, `profileID`, `internalWebPropertyID` constants which might be used in the files above.
- Remove unused `applyEntityToReportPath` function and associated test.

### Test Coverage

* No new tests needed.
* Remove test for `applyEntityToReportPath`

### Visual Regression Changes
- No visual regressions

## QA Brief

* Use a Site Kit install with Analytics to ensure all deep links to Google Analytics website from Site Kit components still work (both with a "default"/main Google account and a secondary Google account that isn't the same as the user's "default" Google account).
* Inspect all "Source links" eg links to Analytics, Search Console, etc. in the plugin and ensure they still function properly after this change with no errors encountered on the "Google" version of these sites.

## Changelog entry

* Update all Analytics deep links to use the `getServiceReportURL` selector for correct encoding of parameters and consistent behavior.

**Comment by aaemnnosttv** [comment]
@asvinb – one thing which should have probably been included in the ACs here is that the `applyEntityToReportPath` will no longer be necessary and is essentially superseded by `getServiceReportURL`. This utility should then be removed as part of the changes here. (cc: @felixarntz)

e.g.
```js
// Before
const sourceURI = useSelect( ( select ) => select( STORE_NAME ).getServiceURL( {
	path: applyEntityToReportPath( url, `/report/trafficsources-overview/a${ accountID }w${ internalWebPropertyID }p${ profileID }/` ),
} ) );

// After
const sourceURI = useSelect(
	( select ) => select( STORE_NAME ).getReportServiceURL( 'trafficsources-overview', { '_r.drilldown': `analytics.pagePath:${ url }` } )
);
```

Another thing to note is that the replacement of `replace( /\//g, '~2F' )` should no longer be done outside of `getReportServiceURL` (`assets/js/modules/analytics/util/report-args.js` to be specific 😄 ). Other than `applyEntityToReportPath`, the only other occurrence of this is in `AnalyticsDashboardWidgetTopPagesTable`.

> - Using `assets/js/modules/analytics/components/dashboard/AnalyticsDashboardWidget.js`
> 	- Update the `topContentServiceURL`, `topAcquisitionServiceURL`, `visitorsOverview` variables to use the `getServiceReportURL` selector:
> 		```js
> 		topContentServiceURL: select( STORE_NAME ).getServiceReportURL( 'content-pages' ),
> 		topAcquisitionServiceURL: select( STORE_NAME ).getServiceReportURL( 'trafficsources-overview' ),
> 		visitorsOverview: select( STORE_NAME ).getServiceReportURL( 'visitors-overview' ),
> 		```

Nothing really wrong with the selector use here, I just wanted to point out that the `useSelect` here should be updated to use individual `useSelect` calls for each URL rather than the kind of compound/bulk select which we have decided against using and to prefer using `useSelect` to select single values as a general rule.

> - Using `assets/js/modules/analytics/components/dashboard/AnalyticsDashboardWidgetTopPagesTable.js`
> 	- Update `baseServiceURL` to the following:
> 		```js
> 		const baseServiceURL = useSelect( ( select ) => select( STORE_NAME ).getServiceReportURL( 'content-drilldown', {
> 			'explorer-table.plotKeys': '[]',
> 			'_r.drilldown': `analytics.pagePath`,
> 		} ) );

The service URL used for reports should no longer be concatenated or appended to as is done in this component. For this component specifically (I didn't see any of the others that do this but there may be) – you can reference how this is done in `assets/js/modules/analytics/components/dashboard/AnalyticsAdSenseDashboardWidgetTopPagesTable.js` which was the first component to use the new selector. Note the use of `options.PrimaryLink` and how that is basically a higher-order `Link` using `withSelect` to pass the `href` through the `getServiceReportURL` selector and its args.

**Comment by asvinb** [comment]
@aaemnnosttv Can you take another look at the IB? thanks!

**Comment by aaemnnosttv** [comment]
Thanks @asvinb this is getting closer but still needs a bit of refinement.

> `assets/js/modules/analytics/components/common/AcquisitionPieChart.js`

It isn't clear where `url` is coming from in this one. While this is present in the component, it's always possible this could change due to changes in another issue that are merged before this one is executed. The IB is both more resilient to change as well as easier to read and understand if it doesn't require crosschecking with the code. In this case, you could add "Where `url` is the current entity URL."

The same applies for a few other instances in the IB where `url` is ambiguous.

> Using assets/js/modules/analytics/components/dashboard/AnalyticsDashboardWidgetTopPagesTable.js
> Update baseServiceURL to the following: `withSelect( ( select ) => { ...`

This is not quite correct. The `withSelect( ... )( Link )` approach is for use with `getDataTableFromData` which _is present in this component_ but you're declaring a `baseServiceURL` variable to be a `Link` component type. This intended for use with the `options.PrimaryLink` option not as a URL string. 

> Update the following line:
>	```
>	links[ i ] = `${ baseServiceURL }:${ encodeURIComponent( url.replace( /\//g, '~2F' ) ) }`;
>	```
>	to use the `reportArgsToURLSegment` function which is found in `assets/js/modules/analytics/util/report-args.js`

`reportArgsToURLSegment` shouldn't really be called outside of the `getServiceReportURL` selector. In this case, the individual link in the array probably just needs to be `url` as this will be provided to the `PrimaryLink` component where this link will be provided to the service report selector, the same as it is in `analytics/components/dashboard/AnalyticsAdSenseDashboardWidgetTopPagesTable.js` – we shouldn't be doing any manual concatenation of the service URL.

> assets/js/modules/analytics/components/dashboard/LegacyAnalyticsDashboardWidgetTopLevel.js
> <code>getServiceReportURL( \`visitors-overview/${ pathIDSegment }\`, ... ) </code>

The first argument to `getServiceReportURL` should be a simple URL path segment/slug – in this case `visitors-overview`. In the referenced component, `pathIDSegment` is the segment of the path that contains the account, property, and profile IDs which is already handled/provided by the new selector so this would simply be removed.

---

Overall, it looks like we're on the right track here, with just a few things to work out. One thing I would ask is to try to keep the IB a bit more abstract rather than providing code snippets to copy/paste in each case. The exception here is probably the `reportArgs` as these are fairly specific, but as mentioned above, it helps to offer a bit more context as to where things come from in the surrounding code, without simply making the code snippet larger 🙂 

If you have any questions feel free to reach out and we can go through it together :thumbsup:

**Comment by asvinb** [comment]
@aaemnnosttv The IB has been updated as per your latest feedback. Can you check again? 
Sounds good for the code snippet, I'll have that in mind for future IBs.

**Comment by aaemnnosttv** [comment]
IB ✅  👍

**Comment by aaemnnosttv** [comment]
@felixarntz CR ✅

**Comment by wpdarren** [comment]
### QA Update: Fail ❌ 

@asvinb @tofumatt for the `Top content over the last 28 days` widget in the Analytics module, when I click on any of the title links it should redirect me to Google Analytics (GA) for that particular page. 

When the page loads, this error message  appears. 

```
An illegal state has occurred.
The request resulted in an inappropriate state. Reload the page to see whether the problem persists.
Error ID: 3e21fbcd-3d56-4bf2-bbab-537524845ec2
```
![image](https://user-images.githubusercontent.com/73545194/104586178-190d7300-56a0-11eb-81c1-4ac70ebe8cee.png)

An example of the URL is https://analytics.google.com/analytics/web/?authuser=hello%40hivevine.com#/report/content-drilldown/a110825622w205695307p198802416/explorer-table.plotKeys%3D%255B%255D%26_r.drilldown%3Danalytics.pagePath%253A~2Fwhat-is-freelance-work~2F/

I have closed the error and refreshed the page but the issue still exists.

**Comment by aaemnnosttv** [comment]
@tofumatt looks like the segmentArgs are getting double URL encoded here 🤔

**Comment by aaemnnosttv** [comment]
@wpdarren I'm not able to reproduce the issue you've described here; those links work as expected for me. Perhaps you have some stale cached assets in your browser? It shouldn't matter if you're using a production build since those include hashes but development builds no longer do so you need to be a bit more careful in that case. I'm not sure if that's the issue, but would be the first thing I would check.

I tested with the latest develop (production build) locally and on another site using a few different accounts and was unable to replicate what you saw there, or any other broken link from those.

Sending back to QA for another pass. (cc: @cole10up )

**Comment by cole10up** [comment]
### QA ✅

Tested:
**Confirmed deep links functioned properly on the following pages:**

- Dashboard links for Analytics and Search Console
- Analytics Module page
- Single page stats page
- Settings page
- Setup module page
No issues on my end

I did however notice console errors on the Google side source:
![image](https://user-images.githubusercontent.com/57115907/104689105-cc816200-56b6-11eb-8048-ffd77079fab7.png)

All functions properly within Site Kit with the latest release candidate. Sending to approval
